subproject/simple/simple_dragonbox.h

```diff
@@ -0,0 +1,1996 @@
+
+        auto const intermediate = (bd >> 32) + uint32_t(ad) + uint32_t(bc);
+
+        return {ac + (intermediate >> 32) + (ad >> 32) + (bc >> 32),
+                (intermediate << 32) + uint32_t(bd)};
+    }
+
+    // Get high half of the 128-bit result of multiplication of two 64-bit unsigned
+    // integers.
+    constexpr uint64_t umul128_upper64(uint64_t x, uint64_t y) {
+        auto const a = uint32_t(x >> 32);
+        auto const b = uint32_t(x);
+        auto const c = uint32_t(y >> 32);
+        auto const d = uint32_t(y);
+
+        auto const ac = umul64(a, c);
+        auto const bc = umul64(b, c);
+        auto const ad = umul64(a, d);
+        auto const bd = umul64(b, d);
+
+        auto const intermediate = (bd >> 32) + uint32_t(ad) + uint32_t(bc);
+
+        return ac + (intermediate >> 32) + (ad >> 32) + (bc >> 32);
+    }
+
+    // Get upper 128-bits of multiplication of a 64-bit unsigned integer and a 128-bit
+    // unsigned integer.
+    constexpr uint128 umul192_upper128(uint64_t x, uint128 y) {
+        auto r = umul128(x, y.high);
+        r += umul128_upper64(x, y.low);
+        return r;
+    }
+
+    // Get upper 64-bits of multiplication of a 32-bit unsigned integer and a 64-bit
+    // unsigned integer.
+    constexpr uint64_t umul96_upper64(uint32_t x, uint64_t y) {
+        auto const yh = uint32_t(y >> 32);
+        auto const yl = uint32_t(y);
+
+        auto const xyh = umul64(x, yh);
+        auto const xyl = umul64(x, yl);
+
+        return xyh + (xyl >> 32);
+    }
+
+    // Get lower 128-bits of multiplication of a 64-bit unsigned integer and a 128-bit
+    // unsigned integer.
+    constexpr uint128 umul192_lower128(uint64_t x, uint128 y) {
+        auto const high = x * y.high;
+        auto const high_low = umul128(x, y.low);
+        return {(high + high_low.high) & UINT64_C(0xffffffffffffffff), high_low.low};
+    }
+
+    // Get lower 64-bits of multiplication of a 32-bit unsigned integer and a 64-bit
+    // unsigned integer.
+    constexpr uint64_t umul96_lower64(uint32_t x, uint64_t y) {
+        return (x * y) & UINT64_C(0xffffffffffffffff);
+    }
+
+    template <int k, class Int>
+    constexpr Int compute_power(Int a) {
+        static_assert(k >= 0);
+        int e = k;
+        Int p = 1;
+        while (e) {
+            if (e % 2)
+                p *= a;
+            e /= 2;
+            a *= a;
+        }
+        return p;
+    }
+
+    template <class Integer>
+    struct compute_mul_result {
+        Integer integer_part;
+        bool is_integer;
+    };
+
+    struct compute_mul_parity_result {
+        bool parity;
+        bool is_integer;
+    };
+
+    // These classes expose encoding specs of IEEE-754-like floating-point formats.
+    // Currently available formats are IEEE-754 binary32 & IEEE-754 binary64.
+
+    template <class T>
+    struct float_format;
+
+    template <>
+    struct float_format<float> {
+        using carrier_uint = uint32_t;
+        enum {
```

Comment:
I think using `static constexpr` is fine, as I think that construct has been widely used at least since C++11, and even before C++11, people used `static const` instead. Some people may find `enum` more familiar, but I guess that would be a sensible "simplification" only when the implementation is written in C. Furthermore, this implementation already uses `static constexpr` elsewhere.